daemon/endpoint.go

```diff
@@ -177,6 +180,13 @@ func (d *Daemon) createEndpoint(epTemplate *models.EndpointChangeRequest, id str
 		}
 	}
 
+	// In order to capture initial states, we need to increment the
+	// counters before `SetIdentityLabels` since `SetIdentityLabels` can
+	// trigger an endpoint regenerate
+	// and change the state to "waiting-to-regenerate"
+	metrics.EndpointStateCount.
```

Comment:
I'm not sure this is correct. `PUT /endpoint/{id}` can be called on existing endpoints. It is clearly not symmetric. Why can't we solve this entirely in `SetState` and detect the initial state change `nil -> ...` and the final `... -> disconnected` to do proper accounting?

Reply:
@tgraf `Init` identity state needs to be captured in `createEndpoint` I never see this state change in  `SetState` i.e toState is never `Init` and hence my understanding was that this initial state needs to be captured in `createEndpoint`.  CC ing to confirm this: @rlenglet 

Also, we do not strictly follow `SetState` to set endpoint states.. I can see this in case of `Init` state as well as `disconnected` state and `restoring` state.

Reply:
@tgraf When `PUT /endpoint/{id}` is called on an existing endpoint, that existing endpoint is first deleted:
```
		e, err := endpointmanager.Lookup(params.ID)
		if err != nil {
			// Delete endpoint if any operation that occurs during PUT, fails,
			// so that endpoints in an unhealthy state are not left lying around.
			if e != nil {
				h.d.deleteEndpoint(e)
			}
			return api.Error(PutEndpointIDFailedCode, err)
		}
```

I think nevertheless that calling `SetState` instead of directly modifying `e.state` directly is a good thing to do. There are currently 3 places where `Endpoint.state` is set without calling `SetStateLocked`. Those places need to be modified to call `SetStateLocked` instead.

Reply:
The code you quote does not delete the endpoint when it already exists. It handles the case when the endpoint was removed while we are waiting for the endpoint to be built in sync mode.

Reply:
@tgraf Done. Please take a look. Also handled special cases for `disconnected` state and initial `nil->waiting-for-identity` state. Audited and `e.state` is set now only via calling `SetState*()`